### PR TITLE
New Fedimint Docker Downloader

### PR DESCRIPTION
**Pull request by Kodylow**
New Fedimint Downloader, currently points at this branch if you want to try testing it with:

```
curl -sSL https://raw.githubusercontent.com/Kodylow/fedimint/kl/docker-deploy/docker/downloader.sh | bash
```

Works for Fedimint Guardians and Gateways on mutinynet and mainnet (couldn't get the testnet one working but can be a follow on)

image

### CODE REVIEW

File: docker/downloader.sh [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+    TLS_DIR="no-tls"
+    break
+    ;;
+  *)
+    echo "Invalid option. Please type 'yes' or 'no'."
+    ;;
+  esac
+done
+
+DOCKER_COMPOSE_FILE="https://raw.githubusercontent.com/Kodylow/fedimint/kl/docker-deploy/docker/latest/${INSTALL_TYPE}/${TLS_DIR}/docker-compose.yaml"
```
Comment (TonyGiorgio) [inline docker/downloader.sh:176]: nit: reminder to switch it back to fedimint right before merging

File: docker/gateway/tls/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+      - "traefik.http.routers.gatewayd.tls.certresolver=myresolver"
+  ### END_GATEWAYD ###
+
+  ### START_GATEWAY_UI ###
+  gateway-ui:
+    image: fedimintui/gateway-ui:0.3.0
+    environment:
+      - PORT=80
+      - REACT_APP_FM_GATEWAY_API=https://gatewayd.fedimint.my-super-host.com
+      - REACT_APP_FM_GATEWAY_PASSWORD=thereisnosecondbest
```
Comment (TonyGiorgio) [inline docker/gateway/tls/docker-compose.yaml:120]: I think this should be removed, don't want to expose the gateway to the outside world.

File: docker/guardian/tls/docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# Best used if called by 'tls-download-mutinynet.sh' script
+# fedimint.my-super-host.com should be replaced by a domain controlled by you and all DNS below should redirect to the IP of your server.
+# Domains:
+# - fedimint.my-super-host.com
+# - fedimintd.fedimint.my-super-host.com
+# - guardian-ui.fedimint.my-super-host.com
+# - gatewayd.fedimint.my-super-host.com
+# - gateway-ui.fedimint.my-super-host.com
+# - rtl.fedimint.my-super-host.com
```
Comment (TonyGiorgio) [inline docker/guardian/tls/docker-compose.yaml:9]: These can be removed from the comments on the guardian scripts

File: docker/gateway/tls/docker-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,212 @@
+# Best used if called by 'tls-download-mutinynet.sh' script
+# fedimint.my-super-host.com should be replaced by a domain controlled by you and all DNS below should redirect to the IP of your server.
+# Domains:
+# - fedimint.my-super-host.com
+# - fedimintd.fedimint.my-super-host.com
+# - guardian-ui.fedimint.my-super-host.com
```
Comment (TonyGiorgio) [inline docker/gateway/tls/docker-compose.yaml:6]: these can be removed from the gateway scripts, might want to check all of the yaml comments

File: docker/guardian/tls/docker-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    volumes:
+      - "bitcoin_datadir:/root/.bitcoin"
+    restart: always
+    platform: linux/amd64
+  ### END_BITCOIND ###
+
+volumes:
+  letsencrypt_data:
+  fedimintd_data:
+  bitcoin_datadir:
```
Comment (TonyGiorgio) [inline docker/guardian/tls/docker-compose.yaml:93]: One thing I removed from my scripts which you may want to put back was xmpp. I'm unsure if that was intentionally left out of here or not too.

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]

File: docker/tls-download-mutinynet.sh
Comment (TonyGiorgio) [inline docker/tls-download-mutinynet.sh:1]: should this download script be deprecated completely? It seems like `downloader.sh` encompasses mutinynet as well as various version numbers. You could redirect to it if you wanted to support the URL still working.

**Comment by TonyGiorgio** [comment]
Some minor things I have noticed. I like the structure of the new script. I still need to play around with it to double check the various scenarios.

File: docker/downloader.sh (hunk 1 above)
Comment (Kodylow) [reply docker/downloader.sh:176]: tracking ty

File: docker/guardian/tls/docker-compose.yaml (hunk 5 above)
Comment (Kodylow) [reply docker/guardian/tls/docker-compose.yaml:93]: nope we're pulling all xmpp out from these

File: docker/gateway/tls/docker-compose.yaml (hunk 4 above)
Comment (Kodylow) [reply docker/gateway/tls/docker-compose.yaml:6]: removed in next commit

File: docker/tls-download-mutinynet.sh
Comment (Kodylow) [reply docker/tls-download-mutinynet.sh:1]: I was leaving it just until this one gets the okay, will remove all old docker things in follow on PR

File: docker/guardian/tls/docker-compose.yaml (hunk 5 above)
Comment (TonyGiorgio) [reply docker/guardian/tls/docker-compose.yaml:88]: Is chat going to be removed completely in fedimint?

Comment (Kodylow) [reply docker/guardian/tls/docker-compose.yaml:88]: removing here in the open source repo, fedi specifically uses it not a general fedimint thing

Comment (TonyGiorgio) [reply docker/guardian/tls/docker-compose.yaml:88]: I know other mints that still want to operate and communicate through XMPP through the fedi app. 

Removing it here, and keeping XMPP in fedi, is prioritizing secret functionality only known by fedi deployments.

**Comment by TonyGiorgio** [comment]
I ran this and it went into an infinite loop at the beginning.

Screenshot 2024-05-04 at 5 23 56 PM

**Comment by Kodylow** [comment]
are you running it from inside another docker container or something? just retried locally and worked on my mac and nix boxes fine. it's looping for all the configurations to require your input match one of the enum options, if you're running as a background or outside an interactive shell that might be causing it?

**Comment by TonyGiorgio** [comment]
I'm running it in a digital ocean shell

**Comment by Kodylow** [comment]
This one works, might have to remove the loop checking then, piping into bash directly doesn't start interactive:

`curl -sSL https://raw.githubusercontent.com/Kodylow/fedimint/kl/docker-deploy/docker/downloader.sh > downloader.sh && chmod +x downloader.sh && ./downloader.sh`

**Comment by Kodylow** [comment]
actually bash has a -i that works:

```
curl -sSL https://raw.githubusercontent.com/Kodylow/fedimint/kl/docker-deploy/docker/downloader.sh | bash -i
```

**Comment by TonyGiorgio** [comment]
Neither of those work. Doing it from my own terminal instance through ssh doesn't work either.

**Comment by Kodylow** [comment]
@TonyGiorgio sorry I'm an idiot just had to move the echo out of the loop in a couple places, fixed

**Comment by TonyGiorgio** [comment]
right after clicking the first yes

Screenshot 2024-05-04 at 8 43 11 PM

**Comment by Kodylow** [comment]
1 more time, missed a </dev/tty

**Comment by TonyGiorgio** [comment]
After saying yes to the alpha disclaimer

Screenshot 2024-05-04 at 8 51 11 PM

File: docker/downloader.sh [hunk 6]
```diff
@@ -0,0 +1,648 @@
...
+    if [[ -z ${bitcoin_rpc_url[*]} ]]; then
+      echo 'Error: You must set a Bitcoin RPC URL if you configure the guardian with an existing Bitcoin node' >&2
+      exit 1
+    fi
+    echo "Using Bitcoin RPC URL: $bitcoin_rpc_url"
+
+    # Set the bitcoin RPC URL in the docker-compose file
+    sed -i "s|FM_BITCOIN_RPC_URL=https://blockstream.info/api/|FM_BITCOIN_RPC_URL=$bitcoin_rpc_url|" ./docker-compose.yaml
+  fi
+fi
```
Comment (TonyGiorgio) [inline docker/downloader.sh:589]: Going through the esplora flow still launches the bitcoind docker container and starts syncing the chain. I think you should probably strip out or insert the bitcoind part of the docker-compose.yaml file based on the results here.

File: docker/downloader.sh [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+    ESPLORA_URL=${ESPLORA_URL:-$DEFAULT_ESPLORA_URL}
+    echo "Using Esplora URL: $ESPLORA_URL"
+
+    # Set the esplora URL in the docker-compose file
+    sed -i "s|FM_BITCOIN_RPC_URL=https://blockstream.info/api/|FM_BITCOIN_RPC_URL=$ESPLORA_URL|" ./docker-compose.yaml
+
+    # Set the bitcoin network in the docker-compose file if not mutinynet default
+    if [ "$NETWORK_TYPE" != "mutinynet" ]; then
+      sed -i "s|FM_BITCOIN_NETWORK=signet|FM_BITCOIN_NETWORK=$NETWORK_TYPE|" ./docker-compose.yaml
+    fi
```
Comment (TonyGiorgio) [inline docker/downloader.sh:575]: for mainnet, it's `bitcoin` I believe, not `mainnet`. I'm getting this: 

```
root-fedimintd-1  | For more information, try '--help'.
root-fedimintd-1  | Starting fedimintd
root-fedimintd-1  | error: invalid value 'mainnet' for '--network <NETWORK>': Unknown network (type mainnet)
```

File: docker/0.2.2/fedimintd-mutinynet/docker-compose.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
+# Important: 127.0.0.1 should be replaced by the external ip address of the machine
+version: "3"
+
+services:
+  fedimintd:
+    image: fedimint/fedimintd:v0.2.1
```
Comment (m1sterc001guy) [inline docker/0.2.2/fedimintd-mutinynet/docker-compose.yaml:6]: Should this be v0.2.2?

File: docker/0.2.2/full-tls-mainnet/docker-compose.yaml [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    restart: always
+    platform: linux/amd64
+    labels:
+      - "traefik.enable=true"
+      - "traefik.http.services.gatewayd.loadbalancer.server.port=80"
+      - "traefik.http.routers.gatewayd.rule=Host(`gatewayd.fedimint.my-super-host.com`)"
+      - "traefik.http.routers.gatewayd.entrypoints=websecure"
+      - "traefik.http.routers.gatewayd.tls.certresolver=myresolver"
+
+  gateway-ui:
```
Comment (m1sterc001guy) [inline docker/0.2.2/full-tls-mainnet/docker-compose.yaml:103]: `gateway-ui` is in here twice

File: docker/0.2.2/full-tls-mainnet/docker-compose.yaml [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    image: fedimintui/gateway-ui:0.2.1
```
Comment (m1sterc001guy) [inline docker/0.2.2/full-tls-mainnet/docker-compose.yaml:104]: v0.2.2 as well? Or was there an auth issue with that?

File: docker/0.2.2/fedimintd-mutinynet/docker-compose.yaml (hunk 8 above)
Comment (Kodylow) [reply docker/0.2.2/fedimintd-mutinynet/docker-compose.yaml:6]: nope, didn't cut that there weren't any ui changes

File: docker/0.2.2/full-tls-mainnet/docker-compose.yaml (hunk 10 above)
Comment (Kodylow) [reply docker/0.2.2/full-tls-mainnet/docker-compose.yaml:104]: same as prev comment

**Comment by bradleystachurski** [comment]
I verified in a new vm that the setup script works using mutinynet for both the guardian and gateway.

```
curl -sSL https://raw.githubusercontent.com/Kodylow/fedimint/kl/docker-deploy/docker/downloader.sh | bash
```

File: docker/gateway/tls/docker-compose.yaml [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      - gateway_datadir:/gateway_data
+    restart: always
+    platform: linux/amd64
+    labels:
+      - "traefik.enable=true"
+      - "traefik.http.services.gatewayd.loadbalancer.server.port=80"
+      - "traefik.http.routers.gatewayd.rule=Host(`gatewayd.fedimint.my-super-host.com`)"
+      - "traefik.http.routers.gatewayd.entrypoints=websecure"
+      - "traefik.http.routers.gatewayd.tls.certresolver=myresolver"
+  ### END_LDK_GATEWAYD ###
```
Comment (oleonardolima) [inline docker/gateway/tls/docker-compose.yaml:131]: @Kodylow Do we only want the LDK gateway on the TLS setup ? It's missing on the no-TLS setup.

Comment (Kodylow) [reply docker/gateway/tls/docker-compose.yaml:131]: the LDK gateway is experimental rn and only works with an LSP so testnet or mainnet, both require TLS to interact with the LSP . @tvolk131 would LDK gateway work with no tls in some other way I'm not tracking

**Comment by Kodylow** [comment]
From call: closing in favor of new fedimint-docker repo

**Pull request closed without merging**